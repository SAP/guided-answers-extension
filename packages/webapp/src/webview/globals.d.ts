interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
