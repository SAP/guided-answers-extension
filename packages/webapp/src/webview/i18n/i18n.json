{
    "GUIDED_ANSWER_HEADER": "Guide",
    "SEARCH_GUIDED_ANSWERS": "Search Guided Answers",
    "ALL_ANSWERS": "Home",
    "STEP_BACK": "Step back",
    "RESTART": "Restart",
    "GUIDED_ANSWERS": "Guided Answers",
    "NO_ANSWERS_FOUND": "No answers found",
    "PLEASE_MODIFY_SEARCH": "Please modfiy search",
    "THIS_SOLVED_MY_ISSUE": "This solved my issue",
    "THIS_DID_NOT_SOLVED_MY_ISSUE": "This did not solve my issue",
    "THANK_YOU_TEXT": "We are glad to hear that your issue has been resolved and we hope you enjoy using the Guided Answers application.",
    "THANKS": "Thanks!",
    "HOME": "Home",
    "SEND": "Send",
    "CLOSE": "Close",
    "ISSUE_IS_NOT_RESOLVED": "Issue is not resolved",
    "WE_ARE_SORRY_TO_HEAR_THAT_YOUR_ISSUE_IS_NOT_YET_RESOLVED": "We are sorry to hear that your issue is not yet resolved.",
    "PLEASE_TELL_US_IF_THIS_ANSWER_WAS_HELPFUL": "Please tell us if this answer was helpful",
    "THERE_ARE_SEVERAL_OPTIONS_FOR_GETTING_FURTHER_ASSISTANCE": "There are several options for getting further assistance:",
    "SEARCH_FOR_ANOTHER_GUIDED_ANSWER": "Search for another Guided Answer",
    "START_AN_EXPERT_CHAT": "Start an Expert Chat",
    "SCHUEDULE_AN_EXPERT": "Schedule an Expert",
    "OPEN_AN_INCIDENT": "Open an Incident",
    "ASK_THE_SAP_COMMUNITY": "Ask the SAP Community",
    "FEEDBACK": "Feedback",
    "FEEDBACK_DIALOG_TITLE": "Is this content helpful?",
    "FEEDBACK_DIALOG_SUBTEXT": "If you have suggestions on how to improve this content we would love to hear them!",
    "FEEDBACK_DIALOG_DISCLAIMER": "Your feedback is anonymous, we do not collect any personal data.",
    "FEEDBACK_DIALOG_SUGGESTION": "Your suggestion",
    "FEEDBACK_SENT_DIALOG_TITLE": "Message sent",
    "FEEDBACK_SENT_DIALOG_SUBTEXT": "Thanks for the feedback!"
}
