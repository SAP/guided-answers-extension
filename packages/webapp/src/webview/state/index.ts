export { actions, store } from './store';
export { getInitialState } from './reducers';
