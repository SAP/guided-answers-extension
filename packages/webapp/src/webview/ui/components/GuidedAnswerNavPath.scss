* {
    box-sizing: border-box;
    outline: none;
}

h1 span {
    font-weight: 600;
}

.container {
    width: 100%;
    margin-top: 107px;
    position: relative;
    overflow: hidden;
    padding-left: 20px;
    padding-right: 20px;
}

.vertical-line {
    display: block;
    left: 9px;
    width: 1px;
    position: absolute;
    top: 47px;
    margin-left: -1px;
    height: 55px;
    background: lightgray;
    z-index: 1;
}

.timeline-block {
    width: 100%;
    margin: 0px 0px 0px 10px;
    display: flex;
    justify-content: space-between;
    clear: both;
    position: relative;
    float: none;
    height: 80px;
}

.marker {
    position: absolute;
    top: 20px;
    font-size: 20px;
}

.timeline-content {
    width: 100%;
    margin: 10px 10px 10px 35px;
    border: 1px solid lightgray;
    color: #666;
}

.timeline-content-arrow {
    position: absolute;
    top: 16px;
    left: -7px;
    position: relative;
    width: 0.75rem;
    height: 0.75rem;
    background-color: var(--vscode-editor-background);
    opacity: 1;
    -moz-opacity: 1;
    -webkit-opacity: 1;
    border-bottom: 0.0625rem solid #d1d1d1;
    border-left: 0.0625rem solid #d1d1d1;
    transform: translateX(0px) translateY(0px) rotate(45deg);
    -moz-transform: translateX(0px) translateY(0px) rotate(45deg);
    z-index: 1;
}

.timeline__path {
    display: inline-block;
    color: var(--vscode-textLink-foreground);
    cursor: pointer;
    padding: 1px 7px 2px 18px;
}
