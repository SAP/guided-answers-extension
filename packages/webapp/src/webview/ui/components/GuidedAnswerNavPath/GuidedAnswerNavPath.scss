* {
    box-sizing: border-box;
    outline: none;
}

h1 span {
    font-weight: 600;
}

.container {
    width: 25%;
    margin-top: 10px;
    position: fixed;
    overflow: visible;
}

.vertical-line {
    width: 1px;
    margin-left: 50%;
    height: 10px;
    background: var(--vscode-editorWidget-border);
    z-index: -1;
    margin-top: -10px;
}

.timeline-block {
    width: 100%;
    display: flex;
    justify-content: space-between;
    clear: both;
    position: relative;
    float: none;
    flex-direction: column;
}

.timeline-content {
    width: 100%;
    margin: 0px 10px 10px 0px;
    border-radius: 3px;
    color: var(--vscode-foreground);
    display: flex;
    justify-content: center;
    align-items: flex-start;
    flex-direction: column;
    background: var(--vscode-menu-background);
    border: 1px solid var(--vscode-editorWidget-border);
}

.timeline-content:hover {
    cursor: pointer;
    border: 1px solid var(--vscode-focusBorder, #007fd4);
}

.timeline-content-bottom-border {
    border-bottom: 3px solid var(--vscode-focusBorder, #007fd4);
    pointer-events: none;
}

.timeline-content-bottom-border:hover {
    border-bottom: 3px solid var(--vscode-focusBorder, #007fd4);
    pointer-events: none;
}

.timeline-content-bottom-border-not-solved {
    border-bottom: 3px solid var(--vscode-notificationsErrorIcon-foreground, #f14c4c);
    pointer-events: none;
}

.timeline-content-bottom-border-not-solved :hover {
    pointer-events: none;
}

.timeline__path {
    display: inline-block;
    color: var(--vscode-menu-foreground, #cccccc);
    padding: 10px;
    font-weight: 400;
    font-size: 13px;
    line-height: 18px;
    font-style: normal;
}

.timeline-content-title-large {
    display: block;
    text-align: left;
}
.timeline-content-title-small {
    display: none;
    text-align: left;
}

@media only screen and (max-width: 625px) {
    .timeline-content-title-large {
        display: none;
    }
    .timeline-content-title-small {
        display: block;
    }
    .timeline-content {
        align-items: center;
    }
    .vertical-line {
        left: 46%;
    }
    .container {
        width: 40px;
    }
    .timeline__path {
        width: 34px;
        text-align: center;
    }
}
