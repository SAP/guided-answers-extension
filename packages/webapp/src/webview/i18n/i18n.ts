import { use } from 'i18next';
import { initReactI18next } from 'react-i18next';
import * as i18nEn from './i18n.json';

export { i18n } from 'i18next';

export function initI18n(): void {
    use(initReactI18next)
        .init({
            resources: {
                en: {
                    translation: i18nEn
                }
            },
            lng: 'en',
            fallbackLng: 'en',
            interpolation: {
                escapeValue: false
            }
        })
        .catch((error) => console.error(error));
}
