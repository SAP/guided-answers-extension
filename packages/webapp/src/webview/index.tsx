import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
