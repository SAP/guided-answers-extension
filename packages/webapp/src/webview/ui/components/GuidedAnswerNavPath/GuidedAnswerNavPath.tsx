import { GuidedAnswerNode } from '@sap/guided-answers-extension-types';
import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';
import { actions } from '../../../state';
import { AppState } from '../../../types';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';
import './GuidedAnswerNavPath.scss';
import { focusOnElement } from '../utils';

/**
 * Renders and return the navigation section.
 *
 * @returns - react element for navigation section
 */
export function GuidedAnswerNavPath(): ReactElement {
    let firstTimeFocus: boolean;
    const nodes = useSelector<AppState, GuidedAnswerNode[]>((state) => state.activeGuidedAnswerNode);
    const isSolved = useSelector<AppState, boolean | null>((state) => state.guideFeedback);
    let activeNode = nodes[nodes.length - 1];
    let lastIndex = nodes.length - 1;
    let lastBlockBorderStyle =
        isSolved === false ? 'timeline-content-bottom-border-not-solved' : 'timeline-content-bottom-border';

    if (activeNode) {
        firstTimeFocus = true;
        return (
            <nav className="container">
                <FocusZone
                    direction={FocusZoneDirection.vertical}
                    role="tree"
                    onFocus={() => {
                        if (firstTimeFocus) {
                            focusOnElement('.timeline-content');
                            firstTimeFocus = false;
                        }
                    }}
                    isCircularNavigation={true}>
                    {nodes.map((node, i) => {
                        return (
                            <div key={`timeline-block-${i}`} className="timeline-block" role="treeitem">
                                <button
                                    className={`timeline-content ${i === lastIndex ? lastBlockBorderStyle : ''}`}
                                    onClick={(): void => {
                                        actions.updateActiveNode(node);
                                    }}>
                                    <div className="timeline__path" title={node.TITLE}>
                                        <span className="timeline-content-title-small bold-text">{i + 1}</span>
                                        <span className={`timeline-content-title-large`}>{node.TITLE}</span>
                                    </div>
                                </button>
                                {i !== lastIndex ? <div className="vertical-line"></div> : <></>}
                            </div>
                        );
                    })}
                </FocusZone>
            </nav>
        );
    }
    return <></>;
}
