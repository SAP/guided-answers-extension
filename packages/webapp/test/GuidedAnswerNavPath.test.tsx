import React from 'react';
import { shallow } from 'enzyme';
import { GuidedAnswerNavPath } from '../src/webview/ui/components/GuidedAnswerNavPath';
import { initI18n } from '../src/webview/i18n';
import { actions } from '../src/webview/state';

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest
        .fn()
        .mockReturnValue({ activeGuidedAnswerNode: [] })
        .mockReturnValueOnce([
            {
                BODY: '<p>SAP Fiori Tools is a set of extensions for SAP Business Application Studio and Visual Studio Code</p>',
                EDGES: [
                    { LABEL: 'Deployment', TARGET_NODE: 45996, ORD: 1 },
                    { LABEL: 'Fiori Generator', TARGET_NODE: 48363, ORD: 2 }
                ],
                NODE_ID: 45995,
                QUESTION: 'I have a problem with',
                TITLE: 'SAP Fiori Tools'
            }
        ])
}));

jest.mock('../src/webview/state', () => {
    return {
        actions: {
            updateActiveNode: jest.fn()
        }
    };
});

describe('<GuidedAnswerNavPath />', () => {
    let wrapper: any;
    initI18n();
    beforeEach(() => {
        wrapper = shallow(<GuidedAnswerNavPath />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a GuidedAnswerNavPath component', () => {
        const component = wrapper.html();
        expect(component).toMatchInlineSnapshot(
            `"<nav class="container"><div role="tree" class="ms-FocusZone css-101" data-focuszone-id="FocusZone0"><div class="timeline-block" role="treeitem"><button class="timeline-content timeline-content-bottom-border"><div class="timeline__path" title="SAP Fiori Tools"><span class="timeline-content-title-small bold-text">1</span><span class="timeline-content-title-large">SAP Fiori Tools</span></div></button></div></div></nav>"`
        );

        //Test click event
        wrapper.find('.timeline-content').simulate('click');
        expect(actions.updateActiveNode).toBeCalled();
    });

    it('Should render an empty GuidedAnswerNavPath component', () => {
        expect(wrapper.find('Fragment').length).toBe(1);
    });
});
