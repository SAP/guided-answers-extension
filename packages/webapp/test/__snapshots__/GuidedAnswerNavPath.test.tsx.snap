// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuidedAnswerNavPath /> Should render a GuidedAnswerNavPath component: <nav class="container"><div class="timeline-block"><div class="timeline-content timeline-content-bottom-border"><div class="timeline__path" title="SAP Fiori Tools"><span class="timeline-content-title-small bold-text">1</span><span class="timeline-content-title-large bold-text">SAP Fiori Tools</span></div></div></div></nav> 1`] = `"<nav class=\\"container\\"><div class=\\"timeline-block\\"><div class=\\"timeline-content timeline-content-bottom-border\\"><div class=\\"timeline__path\\" title=\\"SAP Fiori Tools\\"><span class=\\"timeline-content-title-small bold-text\\">1</span><span class=\\"timeline-content-title-large bold-text\\">SAP Fiori Tools</span></div></div></div></nav>"`;
